torch/csrc/autograd/engine.cpp

```diff
@@ -885,33 +969,57 @@ bool Engine::is_checkpoint_valid() {
   return checkpoint_valid;
 }
 
-size_t Engine::ready_queue_size(at::Device device) {
-  if (ready_queues_.empty()) {
-    // The vector ready_queues_ is initialized in start_threads, but this method
-    // can be called before start_threads. Adding this check to avoid index
+std::shared_ptr<ReadyQueue> Engine::init_local_ready_queue(std::shared_ptr<ReadyQueue> ready_queue) {
+  if (ready_queue) {
+    // if ready_queue provided in the caller, use the caller's ready_queue to initialize local_ready_queue
+    TORCH_INTERNAL_ASSERT(!local_ready_queue);
+    local_ready_queue = std::move(ready_queue);
+  } else if (!local_ready_queue){
+    // otherwise if local_ready_queue not allocated, allocate a new ready_queue
+    local_ready_queue = std::make_shared<ReadyQueue>();
+  }
+
+  // if local_ready_queue already exist (i.e. from a previous backward call),
+  // we could also reuse it instead of creating a new one.
+  return local_ready_queue;
```

Comment:
Should we clear the `local_ready_queue` before we return it to the user? Is it possible that there are some tasks in the thread_local ready_queue from previous backwards calls that failed?

Reply:
discussed offline, we shouldn't clear the `local_ready_queue`, as it might get used by the reentrant_thread and there might be additional tasks left in this case.